Roadmap 2022 (#23)
ethereum/js-team-organization: roadmap/EF2022.md

# EF JavaScript Team Roadmap 2022

This document reflects our initial planning for the respective year and represents
the roadmap we have submitted internally within the EF.

Things in the Ethereum ecosystem move quickly, so please note that we will likely deviate
substantially from this initial roadmap throughout the year.

See our "live" [roadmap](./README.md) document for an up-to-date overview on what we
are currently up to.

## Tracks

Roadmap is divided into following "tracks", identifying our main focus areas
for the year.

### DevEx & VM

The [EthereumJS](https://github.com/ethereumjs/ethereumjs-monorepo) libraries are often used in a developement context, being it as an integral part of main Ethereum developer tools or as building blocks for browser UIs of decentralized applications (dApps). Improving the experience when working with our libraries is therefore a crucial part of our work with the VM playing some special role here due to its central place it takes in development tools like Hardhat, Truffle or Remix.

### Harforks & EIPs

> Implementing hardforks and providing early-on access to the added or changed functionality in development tools takes a reocurring share in our work. We also do selected implementations of dedicated EIPs not yet "hardfork-settled" where it makes sense.

Reviewer: ```suggestion
Implementing hardforks and providing early-on access to the added or changed functionality in development tools takes a reoccurring share in our work. We also do selected implementations of dedicated EIPs not yet "hardfork-settled" where it makes sense.
```

### Research & Client

The JavaScript team takes part in Ethereum protocol related research especially in areas where possible future applications in a TypeScript/JavaScript/Web Browser context are on the horizon, going "stateless" is a good example here since it is desirable to have lightweight browser access to the Ethereum network. We use our EthereumJS [client implementation](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/client) here as a research leveraging tool and as a testbed. 

### Portal Network

> The [Portal Network](https://github.com/ethereum/portal-network-specs) is a new peer-to-peer network topology for the Ethereum network (like devp2p) being build to allow for lightweight Ethereum (mainnet) protocol access and expose a JSON-RPC like interface along. Effort is lead by [Piper Merriam](https://github.com/pipermerriam) and progress has been substantially advanced with specs laid out on a detailed level and two client implementations (Trin (Rust) and Fluffy (nim)) under way.

Reviewer: ```suggestion
The [Portal Network](https://github.com/ethereum/portal-network-specs) is a new peer-to-peer network topology for the Ethereum network (like devp2p) being built to allow for lightweight Ethereum (mainnet) protocol access and expose a JSON-RPC like interface along. The effort is led by [Piper Merriam](https://github.com/pipermerriam) and progress has been substantially advanced with specs laid out on a detailed level and two client implementations (Trin [Rust] and Fluffy [Nim]) underway.
```

The EF JavaScript team has build out a [prototype](https://github.com/acolytec3/portalnetwork) on this during 2021 to test on feasability. We will exand on this in 2022 with dedicated hires to build up a client in TypeScript/JavaScript which will allow to bring portal network functionality to the browser.

### Ethers.js

The [Ethers.js](https://github.com/ethers-io/ethers.js) library is one of the popular connector and wallet libraries to get access to and interact with the Ethereum network. Primarily maintained by [Richard Moore](https://github.com/ricmoo) the library is downloaded by far over 1 million users and developers each month. To further develop and maintain the library are therefore crucial to support the large community which has build around the library over the years.

## Q1

### DevEx & VM

- [ ] Continued L2 Support Improvements (Polygon, Optimism, Arbitrum,...)
- [ ] ZK Rollup L2 Solution Support Analysis

> - [ ] VM Support Feasability Analysis for EVM equivalent L2 Solutions (e.g. Optimism)

Reviewer: ```suggestion
- [ ] VM Support Feasibility Analysis for EVM equivalent L2 Solutions (e.g. Optimism)
```

- [ ] Draft implementation for 1-2 sub networks (e.g. state / history)
- [ ] Analysis on browser networking topology
- [ ] Small-scale browser syncing/communication experiments
- [ ] 1-2 additional hires for the project

### Ethers.js

TODO

## Q2

### DevEx & VM

- [ ] Better ZK Rollup L2 Support (depending on analysis from Q1)
- [ ] VM Support for selected EVM equivalent L2 Solution (e.g. Optimism, depending on analysis from Q1)
- [ ] Deeper reaching VM performance optimizations Part II (Memory optimizations, Buffer, Algorithms)

### Harforks & EIPs

- [ ] Iterations on the Merge HF library releases
- [ ] Eventually dedicated work on Merge HF improvement EIPs
- [ ] Continuation on Shanghai HF EIP implementations (scope still open)

### Research & Client

- [ ] First prototype native JavaScript Verkle Tree implementation (depends on Q1 hire)

### Portal Network

- [ ] Stable communication with other clients for sub networks from Q1
- [ ] Draft implementation for 1-2 additional sub networks (e.g. tx / header gossip)
- [ ] Final decisions on browser networking topology
- [ ] Mid-scale browser syncing/communication experiments, feasability study

> 

Reviewer 2: ```suggestion
- [ ] Establish partnership with one or more dapp/wallet providers to build a proof of concept integration that leverages Portal Network data in a real-world setting (e.g. replacing Infura with the Portal Network State Subnetwork for retrieving account balance data)
```

### Ethers.js

TODO

## Q3

### DevEx & VM

- [ ] Integration of yet to be named L2 solutions
- [ ] Deeper reaching VM performance optimizations Part III (Memory optimizations, Buffer, Algorithms)

### Harforks & EIPs

- [ ] Eventually dedicated work on Merge HF improvement EIPs
- [ ] Shanghai or post-Shanghai EIP work

### Research & Client

- [ ] Prototype [stateless roadmap](https://notes.ethereum.org/Yn_mwNa2SeeQHnKsRgekKg) integration (State Expiry)
- [ ] Production-ready Verkle Tree implementation

### Portal Network

> - [ ] Functionally complete production prototype (alpha)

Reviewer 2: ```suggestion
- [ ] Functionally complete integration with a dapp or wallet provider (alpha)
```
